### ISSUE TITLE
Very high memory usage/leak in seafile-applet

### ISSUE THREAD

**Issue by DaSpawn** [haiwen/seafile#1006]
I have 4 different seafile servers my client connects to with about 25+ libraries and 25G+ of data between them.  I am running seafile client 4.0.4 on Ubuntu 14.04.1 LTS x64, and the seafile servers are various mostly newer versions.  I came back from vacation and system memory usage was very high (system has been running for about 1 month), and seafile-applet is using 7.4G of memory!  (luckily I have enough on my machine).  After closing seafile-applet and opening it back up, the memory usage is now 13.3 MB!   My machine is normally running for months without restarts and I have never seen something like this before..

So is this a horrible memory leak, or something else?

**Comment by killing** [comment]
@Chilledheart Please have a look at this.

**Comment by Chilledheart** [comment]
got it

**Comment by bill-mcgonigle** [comment]
I'm willing to test patches.   My first attempt at seafile (4.0.5, Fedora 20) and seaf-daemon is using 12.4GB of resident RAM and 27GB of virtual.  The whole folder it's trying to sync is only 11GB.

![seaf-daemon-memory](https://cloud.githubusercontent.com/assets/880256/5650835/00c97e70-9670-11e4-9cbe-40723933d747.png)

**Comment by bill-mcgonigle** [comment]
I think I might have found my problem, a failure to break recursive symlinks.  In my folder was a udev source tree. 

strace shows a neverending stream of these:

```
[pid 15431] inotify_add_watch(26, "/home/flowerpt/Download/gxv3140/user/udev-094/test/sys/devices/pci0000:00/0000:00:09.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:09.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:09.0/bus/devices/0000:00:09.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:09.0/bus/devices/0000:00:09.0/bus/devices/0000:00:09.0/bus/devices/0000:00:09.0/bus/devices/0000:00:09.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:1e.0/bus/devices/0000:00:09.0/bus/drivers", IN_MODIFY|IN_ATTRIB|IN_CLOSE_WRITE|IN_MOVE|IN_CREATE|IN_DELETE) = 5214
```

this source tree has many symlinks, such as:

```
$ find ./ -type l -ls 
655799    1 lrwxrwxrwx   1 flowerpt flowerpt       56 Mar 28  2012 ./test/sys/bus/pci/drivers/aic7xxx/0000:02:05.0 -> ../../../../devices/pci0000:00/0000:00:1e.0/0000:02:05.0
655800    1 lrwxrwxrwx   1 flowerpt flowerpt       53 Mar 28  2012 ./test/sys/bus/pci/devices/0000:02:05.0 -> ../../../devices/pci0000:00/0000:00:1e.0/0000:02:05.0
655801    1 lrwxrwxrwx   1 flowerpt flowerpt       40 Mar 28  2012 ./test/sys/bus/pci/devices/0000:00:1e.0 -> ../../../devices/pci0000:00/0000:00:1e.0
655802    1 lrwxrwxrwx   1 flowerpt flowerpt       40 Mar 28  2012 ./test/sys/bus/pci/devices/0000:00:09.0 -> ../../../devices/pci0000:00/0000:00:09.0
655798    1 lrwxrwxrwx   1 flowerpt flowerpt       79 Mar 28  2012 ./test/sys/bus/scsi/devices/0:0:0:0 -> ../../../devices/pci0000:00/0000:00:1e.0/0000:02:05.0/host0/target0:0:0/0:0:0:0
655797    1 lrwxrwxrwx   1 flowerpt flowerpt       82 Mar 28  2012 ./test/sys/bus/scsi/drivers/sd/0:0:0:0 -> ../../../../devices/pci0000:00/0000:00:1e.0/0000:02:05.0/host0/target0:0:0/0:0:0:0
655805    1 lrwxrwxrwx   1 flowerpt flowerpt       48 Mar 28  2012 ./test/sys/bus/usb/drivers/usb/usb3 -> ../../../../devices/pci0000:00/0000:00:09.0/usb3
655806    1 lrwxrwxrwx   1 flowerpt flowerpt       52 Mar 28  2012 ./test/sys/bus/usb/drivers/usb/3-1 -> ../../../../devices/pci0000:00/0000:00:09.0/usb3/3-1
655804    1 lrwxrwxrwx   1 flowerpt flowerpt       56 Mar 28  2012 ./test/sys/bus/usb/drivers/hub/3-0:1.0 -> ../../../../devices/pci0000:00/0000:00:09.0/usb3/3-0:1.0
655803    1 lrwxrwxrwx   1 flowerpt flowerpt       60 Mar 28  2012 ./test/sys/bus/usb/drivers/pl2303/3-1:1.0 -> ../../../../devices/pci0000:00/0000:00:09.0/usb3/3-1/3-1:1.0
655808    1 lrwxrwxrwx   1 flowerpt flowerpt       49 Mar 28  2012 ./test/sys/bus/usb/devices/3-1 -> ../../../devices/pci0000:00/0000:00:09.0/usb3/3-1
655810    1 lrwxrwxrwx   1 flowerpt flowerpt       53 Mar 28  2012 ./test/sys/bus/usb/devices/3-0:1.0 -> ../../../devices/pci0000:00/0000:00:09.0/usb3/3-0:1.0
655809    1 lrwxrwxrwx   1 flowerpt flowerpt       57 Mar 28  2012 ./test/sys/bus/usb/devices/3-1:1.0 -> ../../../devices/pci0000:00/0000:00:09.0/usb3/3-1/3-1:1.0
655807    1 lrwxrwxrwx   1 flowerpt flowerpt       45 Mar 28  2012 ./test/sys/bus/usb/devices/usb3 -> ../../../devices/pci0000:00/0000:00:09.0/usb3
655796    1 lrwxrwxrwx   1 flowerpt flowerpt       57 Mar 28  2012 ./test/sys/bus/usb-serial/devices/ttyUSB0 -> ../../../devices/pci0000:00/0000:00:09.0/usb3/3-1/ttyUSB0
655795    1 lrwxrwxrwx   1 flowerpt flowerpt       60 Mar 28  2012 ./test/sys/bus/usb-serial/drivers/PL-2303/ttyUSB0 -> ../../../../devices/pci0000:00/0000:00:09.0/usb3/3-1/ttyUSB0
655792    1 lrwxrwxrwx   1 flowerpt flowerpt       19 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/bus -> ../../../../bus/usb
655791    1 lrwxrwxrwx   1 flowerpt flowerpt       34 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/3-0:1.0/driver -> ../../../../../bus/usb/drivers/hub
655790    1 lrwxrwxrwx   1 flowerpt flowerpt       22 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/3-0:1.0/bus -> ../../../../../bus/usb
655793    1 lrwxrwxrwx   1 flowerpt flowerpt       31 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/driver -> ../../../../bus/usb/drivers/usb
655786    1 lrwxrwxrwx   1 flowerpt flowerpt       25 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/3-1/3-1:1.0/bus -> ../../../../../../bus/usb
655787    1 lrwxrwxrwx   1 flowerpt flowerpt       40 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/3-1/3-1:1.0/driver -> ../../../../../../bus/usb/drivers/pl2303
655784    1 lrwxrwxrwx   1 flowerpt flowerpt       32 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/3-1/ttyUSB0/bus -> ../../../../../../bus/usb-serial
655785    1 lrwxrwxrwx   1 flowerpt flowerpt       48 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/3-1/ttyUSB0/driver -> ../../../../../../bus/usb-serial/drivers/PL-2303
655789    1 lrwxrwxrwx   1 flowerpt flowerpt       34 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/3-1/driver -> ../../../../../bus/usb/drivers/usb
655788    1 lrwxrwxrwx   1 flowerpt flowerpt       22 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/usb3/3-1/bus -> ../../../../../bus/usb
655794    1 lrwxrwxrwx   1 flowerpt flowerpt       16 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:09.0/bus -> ../../../bus/pci
655779    1 lrwxrwxrwx   1 flowerpt flowerpt       29 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:1e.0/0000:02:05.0/host0/target0:0:0/0:0:0:0/bus -> ../../../../../../../bus/scsi
655780    1 lrwxrwxrwx   1 flowerpt flowerpt       40 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:1e.0/0000:02:05.0/host0/target0:0:0/0:0:0:0/driver -> ../../../../../../../bus/scsi/drivers/sd
655782    1 lrwxrwxrwx   1 flowerpt flowerpt       35 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:1e.0/0000:02:05.0/driver -> ../../../../bus/pci/drivers/aic7xxx
655781    1 lrwxrwxrwx   1 flowerpt flowerpt       19 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:1e.0/0000:02:05.0/bus -> ../../../../bus/pci
655783    1 lrwxrwxrwx   1 flowerpt flowerpt       16 Mar 28  2012 ./test/sys/devices/pci0000:00/0000:00:1e.0/bus -> ../../../bus/pci
655777    1 lrwxrwxrwx   1 flowerpt flowerpt       76 Mar 28  2012 ./test/sys/block/sda/device -> ../../devices/pci0000:00/0000:00:1e.0/0000:02:05.0/host0/target0:0:0/0:0:0:0
655776    1 lrwxrwxrwx   1 flowerpt flowerpt       76 Mar 28  2012 ./test/sys/block/cciss!c0d0/device -> ../../devices/pci0000:00/0000:00:1e.0/0000:02:05.0/host0/target0:0:0/0:0:0:0
655778    1 lrwxrwxrwx   1 flowerpt flowerpt       76 Mar 28  2012 ./test/sys/block/rd!c0d0/device -> ../../devices/pci0000:00/0000:00:1e.0/0000:02:05.0/host0/target0:0:0/0:0:0:0
655811    1 lrwxrwxrwx   1 flowerpt flowerpt       57 Mar 28  2012 ./test/sys/class/tty/ttyUSB0/device -> ../../../devices/pci0000:00/0000:00:09.0/usb3/3-1/ttyUSB0
```

After removing this directory, seaf-daemon is succeeding.  I suppose this makes this related to issue #130 as it's a misbehavior caused by not handling symlinks as a filetype.  I don't know if this is the same as the problem @DaSpawn is having.

**Comment by DaSpawn** [comment]
Looks like mine may have been a bit different, looking at logs and there is 68466 entries with failed to remove:

seafile.log:[12/09/14 13:28:55] vc-utils.c(203): Failed to remove dir /home/user/Seafile/..........: Directory not empty.

Other than that no more errors...

**Comment by DaSpawn** [comment]
Wanted to update this comment again, this problem is still happening, even with version 4.1.2 (about to try 4.1.3).  I can not leave my seafile client running anymore as it uses all of my memory within a day, and I have 16G of memory on my desktop (and usually only uses ~5G of memory and rarely shut down)

I also have another machine that has the same files synced, and does not have this memory issue.  Both machines are running Ubuntu 14.04, fully updated.  

How can I debug where this memory leak is?

**Comment by Chilledheart** [comment]
Maybe you can try checking the `applet.log` to find if there is something important, or use `strace` to trace the `seafile-applet` program:
```sudo strace -p $(pgrep seafile-applet)```

**Comment by bill-mcgonigle** [comment]
@DaSpawn : can you also check your tree with this to see if you have any circular symlinks?
http://serverfault.com/a/265614/190714

**Comment by DaSpawn** [comment]
Side note, if I run seafile without the applet (`seaf-cli start`) the daemon uses 150MB of memory and never goes much more than that

I did: `find . -follow -printf ""`, found nothing.  The seafile logs show nothing out of normal.  the strace however is very interesting, but I do not know what is causing it.  Over and over (about every second) this messages just keeps saying:

`
recvmsg(6, 0x7ffff8cdd460, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(6, 0x7ffff8cdd440, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(8, 0x7ffff8cdd430, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=7, events=POLLIN}, {fd=8, events=POLLIN}, {fd=14, events=POLLIN}, {fd=6, events=POLLIN}, {fd=15, events=POLLIN}, {fd=21, events=POLLIN}, {fd=23, events=POLLIN}, {fd=27, events=POLLIN}, {fd=26, events=POLLIN}, {fd=30, events=POLLIN}, {fd=34, events=POLLIN}, {fd=29, events=POLLIN}, {fd=43, events=POLLIN}], 14, 0) = 0 (Timeout)
recvmsg(6, 0x7ffff8cdd460, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(6, 0x7ffff8cdd440, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(8, 0x7ffff8cdd430, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=7, events=POLLIN}, {fd=8, events=POLLIN}, {fd=14, events=POLLIN}, {fd=6, events=POLLIN}, {fd=15, events=POLLIN}, {fd=21, events=POLLIN}, {fd=23, events=POLLIN}, {fd=27, events=POLLIN}, {fd=26, events=POLLIN}, {fd=30, events=POLLIN}, {fd=34, events=POLLIN}, {fd=29, events=POLLIN}, {fd=43, events=POLLIN}], 14, 0) = 0 (Timeout)
recvmsg(6, 0x7ffff8cdd460, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(6, 0x7ffff8cdd440, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(8, 0x7ffff8cdd430, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=7, events=POLLIN}, {fd=8, events=POLLIN}, {fd=14, events=POLLIN}, {fd=6, events=POLLIN}, {fd=15, events=POLLIN}, {fd=21, events=POLLIN}, {fd=23, events=POLLIN}, {fd=27, events=POLLIN}, {fd=26, events=POLLIN}, {fd=30, events=POLLIN}, {fd=34, events=POLLIN}, {fd=29, events=POLLIN}, {fd=43, events=POLLIN}], 14, 0) = 0 (Timeout)
recvmsg(6, 0x7ffff8cdd460, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(6, 0x7ffff8cdd440, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(8, 0x7ffff8cdd430, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=7, events=POLLIN}, {fd=8, events=POLLIN}, {fd=14, events=POLLIN}, {fd=6, events=POLLIN}, {fd=15, events=POLLIN}, {fd=21, events=POLLIN}, {fd=23, events=POLLIN}, {fd=27, events=POLLIN}, {fd=26, events=POLLIN}, {fd=30, events=POLLIN}, {fd=34, events=POLLIN}, {fd=29, events=POLLIN}, {fd=43, events=POLLIN}], 14, 0) = 0 (Timeout)
recvmsg(6, 0x7ffff8cdd460, 0)           = -1 EAGAIN (Resource temporarily unavailable)
write(42, "\1\4\0+\0\0\3\351301 CLIENT CALL\n[\"seafil"..., 51) = 51
ioctl(42, FIONREAD, [0])                = 0
read(42, "\1\3\0\30\0\0\3\351", 8)      = 8
ioctl(42, FIONREAD, [24])               = 0
read(42, "311 SERVER RET\n{\"ret\":0}", 24) = 24
write(42, "\1\4\0-\0\0\3\351301 CLIENT CALL\n[\"seafil"..., 53) = 53
ioctl(42, FIONREAD, [32])               = 0
read(42, "\1\3\0\30\0\0\3\351", 8)      = 8
ioctl(42, FIONREAD, [24])               = 0
read(42, "311 SERVER RET\n{\"ret\":0}", 24) = 24
recvmsg(6, 0x7ffff8cdd440, 0)           = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(8, 0x7ffff8cdd430, 0)           = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=7, events=POLLIN}, {fd=8, events=POLLIN}, {fd=14, events=POLLIN}, {fd=6, events=POLLIN}, {fd=15, events=POLLIN}, {fd=21, events=POLLIN}, {fd=23, events=POLLIN}, {fd=27, events=POLLIN}, {fd=26, events=POLLIN}, {fd=30, events=POLLIN}, {fd=34, events=POLLIN}, {fd=29, events=POLLIN}, {fd=43, events=POLLIN}], 14, 469
`

Is the "Resource temporarily unavailable" message caused by a file system problem/error or is this a seafile server error?

**Comment by Chilledheart** [comment]
Yeah, `EAGAIN` is soft error, and okay most of time when programming in event-driven model.

**Comment by DaSpawn** [comment]
looks like I was wrong, the seaf-cli acts the same, came in this morning after running 2 days and using 11GB of memory

almost impossible to use seafile now, can not leave it open at all...  and I have 4 different seafile servers I have to sync with...   now I have to keep battling sync conflicts every day

**Comment by shoeper** [comment]
Are you sure everything is right with your setup?

Did not see any other memory leak reports so far. And on Windows is also no Problems.

**Comment by DaSpawn** [comment]
I do not know what was wrong, I never did solve.  I ended up removing (moved) all of my seafile data and syncing the same libraries and it appears to be working ok now.  I also rsynced to test the folders from previous sync to new synced files and no difference between files and structure, so maybe some type of local db corruption in seafile?  I have Seafile clients running on Linux, Mac, Windows, and Android without any issues, just his one machine that was actually a brand new Ubuntu install a few months ago but the same Seafile data that I have had for 2 years, so maybe one of the older versions of the client messed up the local databases.  I also have 1 more machine that was the first/oldest install and never had this client memory issue, even though it is the same Ubuntu and syncs the same data as the machine with the issue.

TL;DR the only solution in this case was to delete all local seafile data and start over with client setup
